### ISSUE TITLE
Rest API Post Issue

### ISSUE THREAD

**Issue by fordey** [NakedObjectsGroup/NakedObjectsFramework#287]
Hi Guys,
Am investigating NOF 11 RestFul objects for us in exposing RestFul endpoints into the domain objects.  I have downloaded the code from the Git repo and have it up and running (v straightforward), created my sample domain (Claims).  I have exposed some simple services that i have been able to interact with (using POSTMAN) very quickly.  For instance invoking all Claim service on the domain.

![PostMan_allClaims_Results](https://user-images.githubusercontent.com/3354596/110100442-520dbe00-7d9a-11eb-950a-a084094e5bb8.PNG)

This is at a service level and returns domain representative objects at the domain level i wish to interact with via GET/POST/PUT/DELETE etc.

GETS on the domain object are working fine,  as are puts. 
![PostMan_GET_Results](https://user-images.githubusercontent.com/3354596/110101429-6e5e2a80-7d9b-11eb-9fe3-62c585b9ffda.PNG)

![PostMan_PUT_Results](https://user-images.githubusercontent.com/3354596/110101558-9b124200-7d9b-11eb-8cab-3ca97788d8d9.PNG)

**However POST are causing an Issue**
I have followed the guide in relation to creating a POST (p77 - 9. Objects Of Type Resource) in particular posting a domain object via a cut down version of it representation.

![PostMan_POST](https://user-images.githubusercontent.com/3354596/110105558-72407b80-7da0-11eb-905d-9be38390c713.PNG)

Error i am getting back in the warning Header is : 
199 RestfulObjects "Values provided may not be persisted as an object (ensure any derived properties are annotated NotPersisted"

Error i am seeing in the server logs is 
NakedObjects.Facade.Impl.FrameworkFacade: 2021-03-05 10:50:57,952 [67] ERROR NakedObjects.Facade.Impl.FrameworkFacade - Transient Integrity failed for: TEOID#2 bad values: UserName: old hash:  new hash 1234
NakedObjects.Rest.App.Demo.RestfulObjectsController: 2021-03-05 10:50:57,954 [67] WARN  NakedObjects.Rest.App.Demo.RestfulObjectsController - Context: NakedObjects.Rest.App.Demo.RestfulObjectsController.PostPersist (Template.Server) State Template.Model.Claim;
NakedObjects.Facade.BadRequestNOSException: Values provided may not be persisted as an object (ensure any derived properties are annotated NotPersisted
   at NakedObjects.Facade.Impl.FrameworkFacade.ValidateTransientIntegrity(ObjectContext nakedObject, String digest)
   at NakedObjects.Facade.Impl.FrameworkFacade.SetObject(INakedObjectAdapter nakedObject, ArgumentsContextFacade arguments, Boolean save)
   at NakedObjects.Facade.Impl.FrameworkFacade.CreateObject(String typeName, ArgumentsContextFacade arguments, Boolean save)
   at NakedObjects.Facade.Impl.FrameworkFacade.<>c__DisplayClass25_0.<Persist>b__0()
   at NakedObjects.Facade.Impl.FrameworkFacade.MapErrors[T](Func`1 f)
   at NakedObjects.Facade.Impl.FrameworkFacade.Persist(String typeName, ArgumentsContextFacade arguments)
   at NakedObjects.Rest.API.ControllerHelpers.PersistObjectAndValidate(IFrameworkFacade frameworkFacade, String domainType, ArgumentsContextFacade argsContext, RestControlFlags flags)
   at NakedObjects.Rest.RestfulObjectsControllerBase.<>c__DisplayClass104_0.<PostPersist>g__PersistObject|0()
   at NakedObjects.Rest.RestfulObjectsControllerBase.<>c__DisplayClass104_0.<PostPersist>b__1()
   at NakedObjects.Rest.RestfulObjectsControllerBase.InitAndHandleErrors(Func`1 f)
Microsoft.AspNetCore.Mvc.StatusCodeResult: 2021-03-05 10:50:57,955 [67] INFO  Microsoft.AspNetCore.Mvc.StatusCodeResult - Executing HttpStatusCodeResult, setting HTTP status code 400
Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker: 2021-03-05 10:50:57,956 [67] INFO  Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker - Executed action NakedObjects.Rest.App.Demo.RestfulObjectsController.PostPersist (Template.Server) in 8.5361ms
Microsoft.AspNetCore.Hosting.Diagnostics: 2021-03-05 10:50:57,959 [67] INFO  Microsoft.AspNetCore.Hosting.Diagnostics - Request finished in 15.8401ms 400 

It looks like it think i have somehow 'Dirtied' the object i feel.

Any ideas as this will be key to whether i can use the RO for further purposes..

For information here is some of the code, it is pretty much unchanged from the code in the Git sample.

Claim.cs
namespace Template.Model
{
    public class Claim
    {
        //All persisted properties on a domain object must be 'virtual'

        [NakedObjectsIgnore]//Indicates that this property will never be seen in the UI
       public virtual int Id { get; set; }

        public virtual string ClaimantName { get; set; }

        public virtual string ClaimantAddress { get; set; }

        public virtual decimal Amount { get; set; }

        [Title]//This property will be used for the object's title at the top of the view and in a link
        public virtual string ClaimReferenceID { get; set; }

    

    }
}

namespace Template.Model
{
    //This example service acts as both a 'repository' (with methods for
    //retrieving objects from the database) and as a 'factory' i.e. providing
    //one or more methods for creating new object(s) from scratch.
    public class ClaimService
    {
        #region Injected Services
        //An implementation of this interface is injected automatically by the framework
        public IDomainObjectContainer Container { set; protected get; }
        #endregion
        public Claim CreateNewClaim()
        {
            //'Transient' means 'unsaved' -  returned to the user
            //for fields to be filled-in and the object saved.

            var claim = Container.NewTransientInstance<Claim>();
            return claim;
        }

        public IQueryable<Claim> AllClaims()
        {
            //The 'Container' masks all the complexities of 
            //dealing with the database directly.
            return Container.Instances<Claim>();
        }

        public IQueryable<Claim> FindClaimsByName(string name)
        {
            //Filters students to find a match
            return AllClaims().Where(c => c.ClaimantName.ToUpper().Contains(name.ToUpper()));
        }

        public IQueryable<Claim> FindClaimById(string id)
        {
            return AllClaims().Where(c => c.ClaimReferenceID.ToUpper().Contains(id.ToUpper()));
        }
    }

}

**Comment by richardpawson** [comment]
First, let me reassure you that the general capability does work, because we are using it all the time in communication between the Naked Objects Client and Server. That said, I don't do much _direct_ interaction with the RO API any more, so am a bit rusty.

Yes, you appear to be using the correct format for the body of your post as stated in the RO spec. However...

I just used Fiddler to inspect the Http traffic between Naked Objects client and server, for a very similar scenario (i.e. saving a transient object) and I noticed a difference. The body being posted in my case (different object details, obviously) looks like this:

`{  "description":{"value":"My special"},
     "discountPct":{"value":"10"},
     "type":{"value":"Promotion"},
     "category":{"value":"Customer"},
     "minQty":{"value":"10"},
     "startDate":{"value":"2021-03-06"},
     "endDate":{"value":"2021-04-05"},
     "x-ro-inline-property-details":false
}`

In other words, structurally similar to yours _except for the absense of the surrounding `members `property_.

Now why this should be different I don't know, because we certainly aimed to implement the spec 100% faithfully (though we have since _added_ some small allowable options).  But I'm not going to speculate until I've spoken to Stef about it.

But, in the meantime, could you just try posting the body without the surrounding `members` property, and report back on whether that then works or not?

**Comment by scascarini** [comment]
Take a look at "Adding further security measures" on p116 of the developer manual. That should explain what's going on. In summary you'll need to include an etag value in the message. But you can plug in whatever hash algorithm you want including ones that generate constants to basically bypass the security.

**Comment by fordey** [comment]
Hi Guys,
Thanks for getting back so quick, got that working :-)

Paul

**Comment by richardpawson** [comment]
Stef: I am re-opening this as an issue with the request just to slightly improve the error message. Can we detect that the Etag is in fact missing (rather than wrong) and just give an error message to that effect. 

(I was going to request that we return an appropriate Http error code instead of throwing a server error - but am persuaded that this is perhaps not appropriate, given that a missing etag might be evidence of a man-in-the-middle attack.)

**Comment by richardpawson** [comment]
@Fordey  - If you see this, can you contact me by email, please?  rpawson at nakedobjects org
